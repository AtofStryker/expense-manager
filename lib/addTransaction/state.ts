import { DEFAULT_CURRENCY } from '../shared/currencies'
import { ObjectOf } from '../types'

export enum TransactionTypes {
  fromUser = 'fromUser',
  imported = 'imported',
  repeated = 'repeated',
}

export type TransactionType = keyof typeof TransactionTypes

export interface Tag {
  id: string
  name: string
  uid: string
}

export enum RepeatingOptions {
  none = 'none',
  inactive = 'inactive',
  daily = 'daily',
  weekly = 'weekly',
  monthly = 'monthly',
  annually = 'annually',
}

export type RepeatingOption = keyof typeof RepeatingOptions

export interface BaseTransaction {
  transactionType: TransactionType
  // NOTE: order might be important
  tagIds: string[]
  currency: string
  isExpense: boolean
  note: string
  repeating: RepeatingOption
}

export interface Transaction extends BaseTransaction {
  id: string
  dateTime: Date
  amount: number
  uid: string
}

export interface AddTransaction extends BaseTransaction {
  useCurrentTime: boolean
  newTags: ObjectOf<Tag>
  dateTime?: Date
  tagInputValue: string
  amount: string
  shouldValidateAmount: boolean
}

export const createDefaultAddTransactionState = (): AddTransaction => ({
  transactionType: TransactionTypes.fromUser,
  amount: '',
  tagIds: [],
  newTags: {},
  currency: DEFAULT_CURRENCY.value,
  tagInputValue: '',
  isExpense: true,
  note: '',
  dateTime: undefined,
  useCurrentTime: true,
  shouldValidateAmount: false,
  repeating: RepeatingOptions.none,
})
