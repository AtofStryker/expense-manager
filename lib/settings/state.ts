import { DEFAULT_CURRENCY, Currency } from '../shared/currencies'
import { FirebaseField } from '../types'

export interface Settings {
  mainCurrency: Currency
  defaultCurrency: Currency
}
export interface ExchangeRates {
  rates: { [key in Currency]: number }
  base: Currency
  date: string
}

export interface Profile extends FirebaseField {
  settings: Settings
  exchangeRates: ExchangeRates
}

export const getInitialState = (user: firebase.User): Profile => ({
  uid: user.uid,
  id: user.uid,
  settings: {
    mainCurrency: DEFAULT_CURRENCY,
    defaultCurrency: DEFAULT_CURRENCY,
  },
  exchangeRates: {
    rates: {
      EUR: 1,
      CHF: 1.0784,
      HKD: 9.0882,
      SEK: 10.2958,
      USD: 1.1726,
      GBP: 0.90013,
      HUF: 345.72,
      CZK: 26.319,
      AUD: 1.6508,
      HRK: 7.4755,
      PLN: 4.4201,
    },
    base: 'EUR',
    date: '2020-08-03',
  },
})
