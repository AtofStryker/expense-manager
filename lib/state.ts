import {
  createDefaultAddTransactionState,
  AddTransaction,
  Tag,
  Transaction
} from './addTransaction/state'
import { ObjectOf } from './types'

export type ScreenTitle =
  | 'add'
  | 'transactions'
  | 'charts'
  | 'tags'
  | 'settings'

export type SignInStatus = 'loggedIn' | 'loggingIn' | 'loggedOut' | 'unknown'

export interface State {
  cnt: number
  // use firebase.auth().currentUser to get the current user
  signInStatus: SignInStatus
  currentScreen: ScreenTitle
  addTransaction: AddTransaction
  tags: ObjectOf<Tag>
  transactions: ObjectOf<Transaction>
}

const state: State = {
  cnt: 0,
  signInStatus: 'unknown',
  currentScreen: 'add',
  addTransaction: createDefaultAddTransactionState([]),
  tags: {},
  transactions: {},
}

export const getInitialState = () => state
