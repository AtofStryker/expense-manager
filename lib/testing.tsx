import React from 'react'

import { LocalizationProvider } from '@material-ui/pickers'
import DateFnsUtils from '@material-ui/pickers/adapter/date-fns'
import { ThemeProvider } from '@material-ui/styles'
import firebasemock from 'firebase-mock'
import { Provider as ReduxProvider } from 'react-redux'
import { Store, applyMiddleware, createStore } from 'redux'
import thunk from 'redux-thunk'

import { TestApi } from './api'
import rootReducer from './redux/rootReducer'
import { State, getInitialState } from './state'
import theme from './theme'

export const reduxify = (Component: React.FC, store: Store) => {
  return (
    <ReduxProvider store={store}>
      <ThemeProvider theme={theme}>
        <LocalizationProvider dateAdapter={DateFnsUtils}>
          <Component />
        </LocalizationProvider>
      </ThemeProvider>
    </ReduxProvider>
  )
}
// TODO: maybe move this to the files alongside real impl.
export const configureTestStore = (
  state: Partial<State> = getInitialState(),
) => {
  const logger = { log: () => null }
  const thunkExtra = {
    logger,
    api: new TestApi(logger),
  }
  const middlewares = [thunk.withExtraArgument(thunkExtra)]
  const store = createStore<State, any, unknown, unknown>(
    rootReducer as any,
    state as any,
    applyMiddleware(...middlewares),
  )

  return store
}

export const byAriaLabel = (value: string) => `[aria-label="${value}"]`

type ExtendedGlobal = NodeJS.Global & { mockFirebase: firebase.app.App }

export const initializeMockFirebase = () => {
  const mockauth = new firebasemock.MockAuthentication()
  mockauth.currentUser = { uid: 'mockedUserId' }
  mockauth.autoFlush()
  const mockfirestore = new firebasemock.MockFirestore()
  mockfirestore.autoFlush()
  const mockSdk = new firebasemock.MockFirebaseSdk(
    null, // realtime database
    () => mockauth,
    () => mockfirestore,
    null, // storage
    null, // messaging
  )
  const firebase = mockSdk.initializeApp()
  ;((global as any) as ExtendedGlobal).mockFirebase = firebase
  return firebase
}

export const getMockedFirebase = () => {
  const g = (global as any) as ExtendedGlobal
  if (!g.mockFirebase) {
    throw new Error(
      'Mock firebase is not present. Be sure to mock it first using "initializeMockFirebase"',
    )
  } else {
    return g.mockFirebase
  }
}
