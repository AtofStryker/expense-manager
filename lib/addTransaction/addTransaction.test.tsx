import { mount } from 'enzyme'
import mockdate from 'mockdate'

import { getInitialState } from '../state'
import { byAriaLabel, configureTestStore, reduxify } from '../testing'

import AddTransaction from './index'

describe('add transaction', () => {
  test('shows loading page', () => {
    const comp = mount(
      reduxify(AddTransaction, configureTestStore(getInitialState())),
    )
    expect(comp.html()).toMatchSnapshot()
  })

  describe('logged in', () => {
    let mockedDate
    let store: ReturnType<typeof configureTestStore>
    let comp: ReturnType<typeof mount>

    beforeEach(() => {
      store = configureTestStore({
        ...getInitialState(),
        signInStatus: 'loggedIn',
      })
      comp = mount(reduxify(AddTransaction, store))
      mockedDate = new Date(2019, 10, 19)
      mockdate.set(mockedDate)
    })

    afterEach(() => {
      mockdate.reset()
    })

    test('shows add transaction page', () => {
      comp = mount(reduxify(AddTransaction, store))
      expect(comp.html()).toMatchSnapshot()
    })

    describe('amount validation', () => {
      const setAmountValidationAndMatchSnapshot = (amount: string) => {
        comp
          .find(byAriaLabel('amount'))
          .find('input')
          .simulate('change', { target: { value: amount } })

        expect(
          comp
            .find(byAriaLabel('amount'))
            .find('input')
            .prop('value'),
        ).toBe(amount)
        expect(store.getState().addTransaction.amount).toBe(amount)
        expect(
          comp
            .find(byAriaLabel('amount'))
            .hostNodes()
            .html(),
        ).toMatchSnapshot()
        expect(store.getState().addTransaction.shouldValidateAmount).toBe(true)
      }

      test('not validated initially', () => {
        expect(
          comp
            .find(byAriaLabel('amount'))
            .hostNodes()
            .html(),
        ).toMatchSnapshot()
        expect(store.getState().addTransaction.shouldValidateAmount).toBe(
          false,
        )
      })

      test('allows floats with 2 decimal places', () => {
        setAmountValidationAndMatchSnapshot('12')
        setAmountValidationAndMatchSnapshot('12.3')
        setAmountValidationAndMatchSnapshot('12.34')
        setAmountValidationAndMatchSnapshot('0')

        setAmountValidationAndMatchSnapshot('13.')
        setAmountValidationAndMatchSnapshot('.45')
        setAmountValidationAndMatchSnapshot('-1')
        setAmountValidationAndMatchSnapshot('-13.3')
      })

      test("doesn't allow to submit with invalid amount", () => {
        comp
          .find(byAriaLabel('add transaction'))
          .hostNodes()
          .simulate('click')

        const state = store.getState()
        expect(state.addTransaction.shouldValidateAmount).toBe(true)
        expect(Object.keys(state.transactions)).toHaveLength(0)
      })
    })
  })
})
