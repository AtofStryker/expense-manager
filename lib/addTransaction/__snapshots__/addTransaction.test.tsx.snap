// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add transaction logged in amount validation not validated initially 1`] = `"<div class=\\"MuiFormControl-root\\" aria-label=\\"amount\\"><label class=\\"MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated\\" data-shrink=\\"false\\" for=\\"amount-id\\">Transaction amount</label><div class=\\"MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd\\"><input aria-invalid=\\"false\\" id=\\"amount-id\\" placeholder=\\"0.00\\" type=\\"text\\" class=\\"MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd\\" value=\\"\\"><div class=\\"MuiInputAdornment-root MuiInputAdornment-positionEnd\\"><svg class=\\"MuiSvgIcon-root MuiSvgIcon-colorPrimary\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\" role=\\"presentation\\" style=\\"visibility: hidden;\\"><path d=\\"M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm5 13.59L15.59 17 12 13.41 8.41 17 7 15.59 10.59 12 7 8.41 8.41 7 12 10.59 15.59 7 17 8.41 13.41 12 17 15.59z\\"></path></svg><svg class=\\"MuiSvgIcon-root MuiSvgIcon-colorPrimary\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\" role=\\"presentation\\"><path d=\\"M18 4H6v2l6.5 6L6 18v2h12v-3h-7l5-5-5-5h7z\\"></path></svg></div></div></div>"`;

exports[`add transaction logged in shows add transaction page 1`] = `"<div class=\\"MuiGrid-root makeStyles-root-1 MuiGrid-container MuiGrid-direction-xs-column MuiGrid-align-items-xs-center MuiGrid-justify-xs-center\\"><div class=\\"MuiPaper-root makeStyles-paper-7 MuiPaper-elevation1 MuiPaper-rounded\\"><div class=\\"MuiGrid-root makeStyles-row-4 MuiGrid-container\\"><div role=\\"group\\" class=\\"MuiButtonGroup-root MuiButtonGroup-fullWidth MuiButtonGroup-contained\\"><button class=\\"MuiButtonBase-root MuiButton-root MuiButton-contained MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedContained MuiButtonGroup-groupedContainedHorizontal MuiButtonGroup-groupedContained MuiButton-containedPrimary MuiButton-fullWidth\\" tabindex=\\"0\\" type=\\"button\\"><span class=\\"MuiButton-label\\">Expense</span><span class=\\"MuiTouchRipple-root\\"></span></button><button class=\\"MuiButtonBase-root MuiButton-root MuiButton-contained MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedContained MuiButtonGroup-groupedContainedHorizontal MuiButtonGroup-groupedContained MuiButton-fullWidth\\" tabindex=\\"0\\" type=\\"button\\"><span class=\\"MuiButton-label\\">Income</span><span class=\\"MuiTouchRipple-root\\"></span></button></div></div><div class=\\"MuiGrid-root makeStyles-row-4\\"><div class=\\"makeStyles-chipField-2 css-2b097c-container\\"><div class=\\" css-yk16xz-control\\"><div class=\\" css-1hwfws3\\"><div class=\\" css-1wa3eu0-placeholder\\" style=\\"opacity: 1; transition: opacity 260ms;\\">Transaction tags...</div><div class=\\"css-b8ldur-Input\\"><div class=\\"\\" style=\\"display: inline-block;\\"><input autocapitalize=\\"none\\" autocomplete=\\"off\\" autocorrect=\\"off\\" id=\\"react-select-3-input\\" spellcheck=\\"false\\" tabindex=\\"0\\" type=\\"text\\" aria-autocomplete=\\"list\\" style=\\"box-sizing: content-box; width: 2px; border: 0px; font-size: inherit; opacity: 1; outline: 0; padding: 0px;\\" value=\\"\\"><div style=\\"position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-size: inherit; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;\\"></div></div></div></div><div class=\\" css-1hb7zxy-IndicatorsContainer\\"><span class=\\" css-1okebmr-indicatorSeparator\\"></span><div aria-hidden=\\"true\\" class=\\" css-tlfecz-indicatorContainer\\"><svg height=\\"20\\" width=\\"20\\" viewBox=\\"0 0 20 20\\" aria-hidden=\\"true\\" focusable=\\"false\\" class=\\"css-6q0nyr-Svg\\"><path d=\\"M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z\\"></path></svg></div></div></div></div></div><div class=\\"MuiGrid-root makeStyles-row-4 MuiGrid-container\\"><div class=\\"MuiGrid-root makeStyles-amount-5 MuiGrid-item\\"><div class=\\"MuiFormControl-root\\" aria-label=\\"amount\\"><label class=\\"MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated\\" data-shrink=\\"false\\" for=\\"amount-id\\">Transaction amount</label><div class=\\"MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd\\"><input aria-invalid=\\"false\\" id=\\"amount-id\\" placeholder=\\"0.00\\" type=\\"text\\" class=\\"MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd\\" value=\\"\\"><div class=\\"MuiInputAdornment-root MuiInputAdornment-positionEnd\\"><svg class=\\"MuiSvgIcon-root MuiSvgIcon-colorPrimary\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\" role=\\"presentation\\" style=\\"visibility: hidden;\\"><path d=\\"M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm5 13.59L15.59 17 12 13.41 8.41 17 7 15.59 10.59 12 7 8.41 8.41 7 12 10.59 15.59 7 17 8.41 13.41 12 17 15.59z\\"></path></svg><svg class=\\"MuiSvgIcon-root MuiSvgIcon-colorPrimary\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\" role=\\"presentation\\"><path d=\\"M18 4H6v2l6.5 6L6 18v2h12v-3h-7l5-5-5-5h7z\\"></path></svg></div></div></div><div class=\\"MuiFormControl-root MuiTextField-root makeStyles-currency-6\\"><label class=\\"MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled\\" data-shrink=\\"true\\">Currency</label><div class=\\"MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl\\"><div class=\\"MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input\\" tabindex=\\"0\\" role=\\"button\\" aria-labelledby=\\" \\" aria-haspopup=\\"listbox\\">EUR</div><input type=\\"hidden\\" value=\\"EUR\\"><svg class=\\"MuiSvgIcon-root MuiSelect-icon\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\" role=\\"presentation\\"><path d=\\"M7 10l5 5 5-5z\\"></path></svg></div></div></div></div><div class=\\"MuiGrid-root makeStyles-row-4\\" style=\\"justify-content: start;\\"><label class=\\"MuiFormControlLabel-root\\"><span class=\\"MuiSwitch-root\\"><span class=\\"MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-331 MuiSwitch-switchBase MuiSwitch-colorPrimary PrivateSwitchBase-checked-332 Mui-checked\\" aria-disabled=\\"false\\"><span class=\\"MuiIconButton-label\\"><input class=\\"PrivateSwitchBase-input-334 MuiSwitch-input\\" type=\\"checkbox\\" value=\\"\\" checked=\\"\\"><span class=\\"MuiSwitch-thumb\\"></span></span><span class=\\"MuiTouchRipple-root\\"></span></span><span class=\\"MuiSwitch-track\\"></span></span><span class=\\"MuiTypography-root MuiFormControlLabel-label MuiTypography-body1\\">Use current date and time</span></label></div><div class=\\"MuiCollapse-container MuiCollapse-hidden\\" style=\\"min-height: 0px;\\"><div class=\\"MuiCollapse-wrapper\\"><div class=\\"MuiCollapse-wrapperInner\\"><div class=\\"MuiGrid-root makeStyles-row-4\\"><div class=\\"MuiFormControl-root MuiTextField-root\\" format=\\"MMMM do HH:mm\\" style=\\"flex: 1;\\"><label class=\\"MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled\\" data-shrink=\\"true\\">Transaction date</label><div class=\\"MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl\\"><input aria-invalid=\\"false\\" readonly=\\"\\" type=\\"text\\" class=\\"MuiInputBase-input MuiInput-input\\" value=\\"November 19th 00:00\\"></div></div></div></div></div></div><div class=\\"MuiGrid-root makeStyles-row-4\\"><div class=\\"MuiFormControl-root\\" style=\\"flex: 1;\\"><label class=\\"MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled\\" data-shrink=\\"true\\" for=\\"tx-repeating\\">Repeating</label><div class=\\"MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl\\"><div class=\\"MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input\\" tabindex=\\"0\\" role=\\"button\\" aria-labelledby=\\" mui-component-select-repeating\\" aria-haspopup=\\"listbox\\" id=\\"mui-component-select-repeating\\">none</div><input name=\\"repeating\\" type=\\"hidden\\" id=\\"tx-repeating\\" value=\\"none\\"><svg class=\\"MuiSvgIcon-root MuiSelect-icon\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\" role=\\"presentation\\"><path d=\\"M7 10l5 5 5-5z\\"></path></svg></div></div></div><div class=\\"MuiGrid-root makeStyles-row-4\\"><div class=\\"MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth\\"><label class=\\"MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated\\" data-shrink=\\"false\\">Additional note</label><div class=\\"MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl\\"><input aria-invalid=\\"false\\" type=\\"text\\" class=\\"MuiInputBase-input MuiInput-input\\" value=\\"\\"></div></div></div><div class=\\"MuiGrid-root makeStyles-row-4\\"><button class=\\"MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-fullWidth\\" tabindex=\\"0\\" type=\\"button\\" aria-label=\\"add transaction\\"><span class=\\"MuiButton-label\\">Add transaction</span><span class=\\"MuiTouchRipple-root\\"></span></button></div></div><div class=\\"MuiBottomNavigation-root makeStyles-bottomNav-368\\"><button class=\\"MuiButtonBase-root MuiBottomNavigationAction-root Mui-selected\\" tabindex=\\"0\\" type=\\"button\\" style=\\"text-transform: capitalize; min-width: unset; padding: unset;\\"><span class=\\"MuiBottomNavigationAction-wrapper\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\" role=\\"presentation\\"><path d=\\"M13 7h-2v4H7v2h4v4h2v-4h4v-2h-4V7zm-1-5C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z\\"></path></svg><span class=\\"MuiBottomNavigationAction-label Mui-selected\\">add</span></span><span class=\\"MuiTouchRipple-root\\"></span></button><button class=\\"MuiButtonBase-root MuiBottomNavigationAction-root\\" tabindex=\\"0\\" type=\\"button\\" style=\\"text-transform: capitalize; min-width: unset; padding: unset;\\"><span class=\\"MuiBottomNavigationAction-wrapper\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\" role=\\"presentation\\"><path d=\\"M3 13h2v-2H3v2zm0 4h2v-2H3v2zm0-8h2V7H3v2zm4 4h14v-2H7v2zm0 4h14v-2H7v2zM7 7v2h14V7H7z\\"></path></svg><span class=\\"MuiBottomNavigationAction-label\\">transactions</span></span><span class=\\"MuiTouchRipple-root\\"></span></button><button class=\\"MuiButtonBase-root MuiBottomNavigationAction-root\\" tabindex=\\"0\\" type=\\"button\\" style=\\"text-transform: capitalize; min-width: unset; padding: unset;\\"><span class=\\"MuiBottomNavigationAction-wrapper\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\" role=\\"presentation\\"><path d=\\"M5 9.2h3V19H5zM10.6 5h2.8v14h-2.8zm5.6 8H19v6h-2.8z\\"></path></svg><span class=\\"MuiBottomNavigationAction-label\\">charts</span></span><span class=\\"MuiTouchRipple-root\\"></span></button><button class=\\"MuiButtonBase-root MuiBottomNavigationAction-root\\" tabindex=\\"0\\" type=\\"button\\" style=\\"text-transform: capitalize; min-width: unset; padding: unset;\\"><span class=\\"MuiBottomNavigationAction-wrapper\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\" role=\\"presentation\\"><path d=\\"M21.41 11.58l-9-9C12.05 2.22 11.55 2 11 2H4c-1.1 0-2 .9-2 2v7c0 .55.22 1.05.59 1.42l9 9c.36.36.86.58 1.41.58.55 0 1.05-.22 1.41-.59l7-7c.37-.36.59-.86.59-1.41 0-.55-.23-1.06-.59-1.42zM5.5 7C4.67 7 4 6.33 4 5.5S4.67 4 5.5 4 7 4.67 7 5.5 6.33 7 5.5 7z\\"></path></svg><span class=\\"MuiBottomNavigationAction-label\\">tags</span></span><span class=\\"MuiTouchRipple-root\\"></span></button><button class=\\"MuiButtonBase-root MuiBottomNavigationAction-root\\" tabindex=\\"0\\" type=\\"button\\" style=\\"text-transform: capitalize; min-width: unset; padding: unset;\\"><span class=\\"MuiBottomNavigationAction-wrapper\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\" role=\\"presentation\\"><path transform=\\"scale(1.2, 1.2)\\" fill=\\"none\\" d=\\"M0 0h20v20H0V0z\\"></path><path transform=\\"scale(1.2, 1.2)\\" d=\\"M15.95 10.78c.03-.25.05-.51.05-.78s-.02-.53-.06-.78l1.69-1.32c.15-.12.19-.34.1-.51l-1.6-2.77c-.1-.18-.31-.24-.49-.18l-1.99.8c-.42-.32-.86-.58-1.35-.78L12 2.34c-.03-.2-.2-.34-.4-.34H8.4c-.2 0-.36.14-.39.34l-.3 2.12c-.49.2-.94.47-1.35.78l-1.99-.8c-.18-.07-.39 0-.49.18l-1.6 2.77c-.1.18-.06.39.1.51l1.69 1.32c-.04.25-.07.52-.07.78s.02.53.06.78L2.37 12.1c-.15.12-.19.34-.1.51l1.6 2.77c.1.18.31.24.49.18l1.99-.8c.42.32.86.58 1.35.78l.3 2.12c.04.2.2.34.4.34h3.2c.2 0 .37-.14.39-.34l.3-2.12c.49-.2.94-.47 1.35-.78l1.99.8c.18.07.39 0 .49-.18l1.6-2.77c.1-.18.06-.39-.1-.51l-1.67-1.32zM10 13c-1.65 0-3-1.35-3-3s1.35-3 3-3 3 1.35 3 3-1.35 3-3 3z\\"></path></svg><span class=\\"MuiBottomNavigationAction-label\\">settings</span></span><span class=\\"MuiTouchRipple-root\\"></span></button></div></div>"`;

exports[`add transaction shows loading page 1`] = `
"<img src=\\"../static/coin.svg\\" alt=\\"coin\\" style=\\"width: 60vw; margin-top: 20vh;\\"><h3 class=\\"MuiTypography-root MuiTypography-h3 MuiTypography-gutterBottom\\" style=\\"text-align: center; margin-top: 10vh;\\">Loading...</h3><style>
        @keyframes rotation {
          from {
            transform: rotateY(0deg);
          }
          to {
            transform: rotateY(1800deg);
          }
        }

        img {
          animation: rotation 2s infinite ease-in-out;
          display: block;
          margin-left: auto;
          margin-right: auto;
        }
      </style>"
`;
